File: cli/engine/verify/verify.go
// Package verify derives a short, human-comparable verification code from a
// WebRTC connection's DTLS certificate fingerprints.
//
// WebRTC data channels are encrypted with DTLS, but the certificate fingerprints
// are exchanged through the signaling server. A malicious or compromised server
// could therefore swap the fingerprints and man-in-the-middle the "peer to peer"
// connection without either side noticing (RFC 8827). Comparing this code out of
// band (over the phone, in person, over an existing trusted channel) closes that
// gap: if a fingerprint was altered, the two peers compute different codes, so a
// mismatch reveals the tampering. This is the ZRTP / Signal "safety number"
// model applied to Floe.
package verify

import (
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"sort"
	"strings"
)

// Code returns a verification code derived from the two DTLS fingerprints of a
// connection (as they appear in the SDP, e.g. "sha-256 AB:CD:..."). The inputs
// are sorted, so each peer computes the same code from its own view of the
// connection when, and only when, no one altered the fingerprints in transit.
func Code(fpA, fpB string) string {
	pair := []string{normalize(fpA), normalize(fpB)}
	sort.Strings(pair)
	sum := sha256.Sum256([]byte(pair[0] + "\n" + pair[1]))
	n := binary.BigEndian.Uint32(sum[:4]) % 100000000 // 8 decimal digits
	return fmt.Sprintf("%04d %04d", n/10000, n%10000)
}

// normalize makes the fingerprint comparison robust to case/whitespace so both
// peers agree on the input for a given certificate.
func normalize(fp string) string {
	return strings.ToUpper(strings.TrimSpace(fp))
}

File: client/lib/transfer/verify.ts
// Floe connection verification code. Mirrors cli/engine/verify/verify.go and MUST
// stay in sync with it so a browser peer computes the same code as a CLI/desktop
// peer.
//
// WebRTC data channels are encrypted with DTLS, but the certificate fingerprints
// are exchanged through the signaling server, which could swap them and MITM the
// "peer to peer" connection (RFC 8827). Comparing this code out of band reveals a
// mismatch if a fingerprint was altered. This is the ZRTP / Signal "safety
// number" model.

function normalize(fp: string): string {
    return fp.trim().toUpperCase();
}

/**
 * verifyCode derives the shared verification code from the two DTLS fingerprints
 * of a connection (as they appear in the SDP, e.g. "sha-256 AB:CD:..."). The
 * inputs are sorted, so both peers compute the same code unless a fingerprint was
 * altered in transit.
 */
export async function verifyCode(fpA: string, fpB: string): Promise<string> {
    const pair = [normalize(fpA), normalize(fpB)].sort();
    const data = new TextEncoder().encode(pair[0] + '\n' + pair[1]);
    const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', data));
    // Big-endian uint32 of the first 4 bytes, mod 1e8, as "NNNN NNNN".
    const u32 = ((digest[0] << 24) | (digest[1] << 16) | (digest[2] << 8) | digest[3]) >>> 0;
    const n = u32 % 100000000;
    const pad = (x: number) => x.toString().padStart(4, '0');
    return `${pad(Math.floor(n / 10000))} ${pad(n % 10000)}`;
}

/** extractFingerprint returns the value of the first "a=fingerprint:" line in an
 *  SDP (e.g. "sha-256 AB:CD:..."), or "" if absent. */
export function extractFingerprint(sdp?: string | null): string {
    if (!sdp) return '';
    for (const line of sdp.split('\n')) {
        const t = line.trim();
        if (t.startsWith('a=fingerprint:')) return t.slice('a=fingerprint:'.length).trim();
    }
    return '';
}
